cpp/include/raft/label/merge_labels.cuh

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2020-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <math.h>
+#include <limits>
+
+#include <raft/cudart_utils.h>
+#include <raft/linalg/init.h>
+#include <raft/cuda_utils.cuh>
+
+namespace raft {
+namespace label {
+
+/** Note: this is one possible implementation where we represent the label
+ *  equivalence graph implicitly using labels_a, labels_b and mask.
+ *  For an additional cost we can build the graph with edges
+ *  E={(A[i], B[i]) | M[i]=1} and make this step faster */
+template <typename Index_, int TPB_X = 256>
+__global__ void __launch_bounds__(TPB_X)
+  propagate_label_kernel(const Index_* __restrict__ labels_a,
+                         const Index_* __restrict__ labels_b,
+                         Index_* __restrict__ R, const bool* __restrict__ mask,
+                         bool* __restrict__ m, Index_ N) {
+  Index_ tid = threadIdx.x + blockIdx.x * TPB_X;
+  if (tid < N) {
+    if (__ldg((char*)mask + tid)) {
```

Comment:
Can you tell why this is necessary if the pointers are already tagged with `const` and `__restrict__`?

Reply:
This code is copied from cuml. I didn't write it.